README.md
# iompair

Iompair (from [the Irish for to transmit, to vector](https://en.wiktionary.org/wiki/iompair#Irish)) is a collection of
programmes to help working with [vectortiles](https://wiki.openstreetmap.org/wiki/Vector_tiles). It allows you to store pbf vector
tiles in a simple directory hiearchy, and serve them over HTTP.

You can `serve` vector tiles from a directory over TMS URLs, use the `stuffer`
to fill you local cache from an upstream vector tile cache, use `expire` to
expire tiles from an `osm2pgsql` tile expire list, and use `cache` to assist
you developing a tile style to prevent you having to regenerate vector tiles
from scratch

# Compiling

    cargo build

# Usage

## iompair serve

Serves a TileCache directory over HTTP over a port.

    iompair serve --port 9000 --tc-path /path/to/files/ --urlprefix http://example.com/mytiles/

It also supports TileStash (`--ts-path`), and ZXY directory layouts
(`--zxy-path`) (where files are stored `/path/to/file/zoom/X/Y.pbf`)

### Directory Layouts

#### TileCache

#### TileStash

TileStash are of the form 

#### ZYX

No splitting, files are stored in format `Z/X/Y.pbf` (e.g. `mb-util` will
create this)

### TileJSON URLs

The [TileJSON]() url is `/index.json`. It is the contents of the file `index.json`
in the root of the `--tc/ts/zxy-path` (or `metadata.json` if that doesn't
exist.

### Merging multiple vector tiles together

`iompair serve` can serve just one set of vector tiles, or multiple sets.

If you have `--zxy-path /data/tiles/` and the directories `/data/tiles/land/`,
`/data/tiles/points/` and `/data/tiles/roads/` with each directory having a
separate set of vectortiles, then you have 3 prefixes for 3 tilesets (`land`,
`points`, `roads`), and:

 * A request to `/0/0/0.pbf` won't work
 * `/land/0/0/0.pbf` or `/land/index.json` will return the tiles from the
   `land` directory (etc)
 * Double underscore (`__`) can be used to concatinate prefixes.
   `/land__points/0/0/0.pbf` will return the `0/0/0.pbf` from `land` directory,
   and then the `0/0/0.pbf` form `points` will be concatinated on afterwards.
   (`/points__land/...` will have them in the other order). This can be
   extended many times. `/land__points__roads/0/0/0.pbf` will be `land`, then
   `points`, the `roads`.

#### TileJSON

Each concatinated tileset will have a TileJSON showing the layers in the
subparts. `/land__points__roads/index.json` etc. The TileJSON will be the
correctly concatinated JSON of the sub tilesets

### Fetching from upstream

If the `--upstream` argument is given, and a tile is requested which doesn't
exist, then `iompair` will request it from this upstream URL. This flag takes 2
arguments, a `PREFIX` and the corresponding `URL` for the upstream for this PREFIX.

Example:

    iompair serve --port 9000 --zxy-path /data/tiles --upstream land http://example.com/landtiles/ --upstream points http://localhost:8080/

When a `land` tile doesn't exist in `/data/tiles/land/`, it will be downloaded
from the URL `http://example.com/landtiles/$ZOOM/$X/$Y.pbf`, saved locally to
the land directory, and then served up to the client. Likewise for `points`.
TileJSON for the upstream URLs is not supported.

### Post Fetch Command

If you use `--upstream`, you can also specify `--post-fetch-command` (which
takes one argument) after a file is downloaded from upstream, this command will
be executed with the local filename of the newly downloaded tile as the first
and only argument.

You can use this to copy a newly downloaded tile to other servers, or do
whatever you want really.

Example:

    iompair serve --port 9000 --zxy-path /data/tiles --upstream land http://example.com/landtiles/ --upstream points http://localhost:8080/ --post-fetch-command /usr/local/bin/copy_tiles.sh

## iompair expire

Reads all the expire filename in a directory, looking for files, parses out the
TMS tile references and updates the files stored in the vector tile directory
from the upstream, pbf vector tile source.

It will not finish, but wait until there are new files available, sleeping
between runs

    iompair expire --tc-path /path/to/vector/tile/store --upstream http://example.com/tiles/ --expire-path /path/to/osm2pgsql/expired-tiles/

## iompair stuffer

Populates (stuffs) a tilecache laidout directory with tiles from an upstream
vector tile source. Specifiy the number of threads with `-T`.

    iompair stuffer --tc-path /path/to/put/vector/tiles --upstream http://example.com/tiles/ -z 14 -b 35.55 -t 71.6 -l -25.93 -r 48.95 -T 20

# Copyright & Licence

Copyright 2016 Geofabrik GmbH, licenced under the GNU General Public Licence
version 3 (or later).


CHANGELOG.md
<a name="v0.9.0"></a>
## v0.9.0 (2017-07-05)


#### Bug Fixes

*   Correct description in Cargo.toml ([0f2000ba](0f2000ba))
*   Actually show version number using the crate_version ([95e36f14](95e36f14))
*   Fix accidentally double download of all files ([879e7ffc](879e7ffc))

#### Features

*   Access-Control-Allow-Origin for all ([2411e028](2411e028))



<a name="v0.8.0"></a>
## v0.8.0 (2017-04-21)


#### Features

*   Ping all upstreams to ensure they work ([8c33d8f4](8c33d8f4))
*   Another macro that might be useful ([301ad637](301ad637))
*   Add len method to URLPathPrefix to see how many elements there are ([0ac2818f](0ac2818f))
*   Convert a URLPathPrefix back to a Vec of Strings ([dca1f5eb](dca1f5eb))
*   Add tilelist which will output tiles that are missing ([e5f35d6a](e5f35d6a))
* **serve:**
  *  --post-fetch-command allows you run arbitraty commands on new files ([f3e7a5b9](f3e7a5b9))
  *  Can now specify upstream URLs so if the file isn't there, it'll be downloaded ([1958fe5e](1958fe5e))
* **tilelist:**  Add min-zoom & zoom function ([8d255470](8d255470))

#### Bug Fixes

*   Allow 'timeout' param in the tilejson ([5a85672f](5a85672f))



<a name="v0.7.0"></a>
## v0.7.0 (2017-01-11)


#### Bug Fixes

* **serve:**  Correct tiles attribute in TileJSON for prefixed tiles ([c774bdd8](c774bdd8))

#### Features

*   Produce better gzip concatinated files ([447ea521](447ea521))
*   Support concatting tiles together ([21502af0](21502af0))
* **serve:**
  *  Support metadata.json if there's no index.json ([602cde4d](602cde4d))
  *  added --verbose flag to print every URL served ([f2286c56](f2286c56))
  *  Serving is now able to use prefix'ed ([2ffd315c](2ffd315c))
  *  URLs can now be prefixed ([f62cf0fe](f62cf0fe))
  *  Support TileStash safe layout as well as TileCache ([4d3e8d53](4d3e8d53))



<a name="v0.6.0"></a>
## v0.6.0 (2016-08-12)


#### Features

*   Output when finished ([7f8054a6](7f8054a6))

#### Bug Fixes

*   Work if maxzoom not specified ([e84b0603](e84b0603))
* **serve:**  Return an empty 200, not 404, when tile doesn't exist ([a689ff13](a689ff13))



<a name=""></a>
##  (2016-06-10)


#### Bug Fixes

*   Explicity handle more errors ([996107a3](996107a3))
* **cache:**
  *  Set cache control headers ([7b532820](7b532820))
  *  Remove iron, do it all with hyper now ([b52617f9](b52617f9))
* **expire:**
  *  Logic bug which prevented it expiring tiles ([3c165b88](3c165b88))
  *  Silent output if nothing to do ([2b825389](2b825389))
* **utils:**
  *  Retry downloads until we get a success, or max retries ([23c84cd4](23c84cd4))
  *  Set timeout for waiting for a response to 1 day ([c5af2172](c5af2172))

#### Features

* **expire:**
  *  Display message if there's an error refreshing a tile ([73c66f7b](73c66f7b))
  *  Can change wait between checks of expire files ([319fa5b2](319fa5b2))
* **utils:**  Use proper rust errors for download & save ([7ea313f6](7ea313f6))



<a name="v0.4.0"></a>
## v0.4.0 (2016-05-25)


#### Bug Fixes

*   Correctly look at file age ([a7a9d51c](a7a9d51c))
*   Support current version of iter-progress ([b7236fbd](b7236fbd))
* **cache:**
  *  Set content-type header on responses ([5aed8bca](5aed8bca))
  *  Don't panic if you cannot download the tile, print error and continue ([7c71f53b](7c71f53b))
  *  Return file contents after saving ([14830392](14830392))

#### Features

*   Add --files-older-than to stuffer ([ab75f81d](ab75f81d))
* **cache:**
  *  Add cache control headers to response to limit caching ([e83bd472](e83bd472))
  *  Print if there's a cache miss/hit ([24e372d4](24e372d4))
  *  Print error if TileJSON is invalid ([524cf9f9](524cf9f9))
  *  Add --maxzoom option to test overzooming ([bc367ba9](bc367ba9))
* **serve:**  Add --urlprefix to override the URL in tilejson output ([ad54840a](ad54840a))



<a name="v0.3.0"></a>
## v0.3.0 (2016-03-22)


#### Features

*   Rename to iompair ([8b8ff07b](8b8ff07b))



<a name="v0.2.0"></a>
## v0.2.0 (2016-03-22)


#### Bug Fixes

*   Support current slippy-map-tiles API ([4569777e](4569777e))
*   Sleep better ([9a23cb98](9a23cb98))
*   Only save the file if there is a 200 status ([2fec2a71](2fec2a71))

#### Features

*   Add vtile-expire which refreshes tile that expired ([3aa5bdf7](3aa5bdf7))
*   Add -z/--max-zoom option to control TileJSON output ([aaa814a1](aaa814a1))
*   vtile-stuffer can now take bounding box ([a54a4bfb](a54a4bfb))
*   TileJSON output now calculated dynamically ([a32e8800](a32e8800))
* **cache:**  Make a subcommand ([3de4ba5c](3de4ba5c))
* **expire:**  Now a subcommand ([84ee0c4a](84ee0c4a))
* **serve:**
  *  Now a subcommand ([5a31ccb4](5a31ccb4))
  *  Serve up locally saved tilejson ([46f6bb4c](46f6bb4c))
* **stuffer:**
  *  Now a subcommand ([23834d6f](23834d6f))
  *  Add --always-download option ([c0be16fd](c0be16fd))
  *  Add --min-zoom arg - skip some zoom levels ([629622b8](629622b8))
  *  Download the upstream tilejson and save ([4d42dec8](4d42dec8))



<a name=""></a>
##  (2016-03-09)


#### Features

*   Add -z/--max-zoom option (default 14) ([f49be184](f49be184))





src/cache.rs
extern crate hyper;
extern crate clap;
extern crate rustc_serialize;
extern crate slippy_map_tiles;

use std::io::Read;
use std::fs;
use std::path::Path;

use hyper::Server;
use hyper::server::Request;
use hyper::server::Response;
use hyper::header::{ContentType, CacheDirective, CacheControl};
use hyper::mime::{Mime, TopLevel, SubLevel};
use hyper::Client;
use hyper::status::StatusCode;

use clap::ArgMatches;
use rustc_serialize::json;

use slippy_map_tiles::Tile;

use utils::{download_url, save_to_file, parse_url, URL};

macro_rules! try_or_ret {
    ($e:expr, $msg:expr) => ( match $e {
        Ok(e) => e,
        Err(e) => {
            println!("{} {:?}", $msg, e);
            return;
        },
    });
}

pub fn cache(options: &ArgMatches) {

    let port = options.value_of("port").unwrap();
    let upstream_url = options.value_of("upstream_url").unwrap().to_string();
    let tc_path = options.value_of("tc_path").unwrap().to_string();
    let maxzoom = options.value_of("maxzoom").and_then(|x| { x.parse::<u8>().ok() });
    // TODO make tc_path absolute

    // get the tilejson details
    let client = Client::new();

    //let mut router = Router::new();

    // FIXME split this into functions
    // FIXME use the URLs in the tile json for fetching tiles

    // Grab the upstream tilejson, but replace the `tiles` key with the url of this

    // This is the parsed, new value for tiles
    let new_tiles = json::Json::from_str(&format!("[\"http://localhost:{}/{{z}}/{{x}}/{{y}}.pbf\"]", port)).unwrap();

    // TODO return appropriate error from upstream
    let mut result = try_or_ret!(client.get(&format!("{}/index.json", upstream_url)).send(), "Error getting TileJSON from upstream");
    
    // Some back and forth to decode, replace and encode to get the new tilejson string
    let mut tilejson_contents = String::new();
    try_or_ret!(result.read_to_string(&mut tilejson_contents), "Error when trying to read tilejson contents");
    let tilejson_0 = try_or_ret!(json::Json::from_str(&tilejson_contents), "TileJSON at {}/index.json is not a valid JSON file");

    let mut tilejson = try_or_ret!(tilejson_0.as_object().ok_or("ERR"), "Error when trying to read tilejson contents").to_owned();
    tilejson.insert("tiles".to_owned(), new_tiles);

    if let Some(z) = maxzoom {
        tilejson.insert("maxzoom".to_owned(), json::Json::U64(z as u64));
    }

    let new_tilejson_contents: String = try_or_ret!(json::encode(&tilejson), "Error when trying to create tilejson contents");

    fn tilejson_handler(res: Response, tilejson_contents: &str) {
        res.send(tilejson_contents.as_bytes()).unwrap_or_else(|e| {
            println!("Error when trying to send tilejson to client: {:?}", e);
        });
    }

    // Handler for tiles

    fn tile_handler(mut res: Response, tc_path: &str, z: u8, x: u32, y: u32, upstream_url: &str) {
        let tile = match Tile::new(z, x, y) {
            None => {
                *res.status_mut() = hyper::status::StatusCode::BadRequest;
                res.send(format!("Invalid tile number {}/{}/{}", z, x, y).as_bytes()).ok();
                return;
            },
            Some(t) => t,
        };
        let path = format!("{}/{}", tc_path, tile.tc_path("pbf"));
        let this_tile_tc_path = Path::new(&path);

        // This is a stupid bit of hackery to ensure that s is initialised to /something/
        let mut vector_tile_contents: Vec<u8> = Vec::new();
        
        if this_tile_tc_path.exists() {
            let mut file = try_or_err!(fs::File::open(this_tile_tc_path), res, format!("Couldn't open tile {:?}", this_tile_tc_path));
            try_or_err!(file.read_to_end(&mut vector_tile_contents), res, "Error when trying to send vectortile contents to client");
            println!("Cache hit {}/{}/{}", z, x, y);
        } else {
            try_or_err!(download_url(&format!("{}/{}/{}/{}.pbf", upstream_url, z, x, y), 10), res,
                format!("Cache miss {}/{}/{} and error downloading file", z, x, y),
                Ok => {
                    try_or_err!(save_to_file(this_tile_tc_path, &vector_tile_contents), res,
                        "Error when trying to save file to cache",
                        Ok => { println!("Cache miss {}/{}/{} Downloaded and saved in {:?}", z, x, y, this_tile_tc_path); }
                    );
                }
            );
        }

        *res.status_mut() = StatusCode::Ok;
        res.headers_mut().set(CacheControl(vec![CacheDirective::Private, CacheDirective::NoCache, CacheDirective::MaxAge(0)]));
        res.headers_mut().set(ContentType(Mime(TopLevel::Application, SubLevel::Ext("x-protobuf".to_owned()), vec![])));
        res.send(&vector_tile_contents).ok();
    }

    fn base_handler(req: Request, mut res: Response, tc_path: &str, upstream_url: &str, tilejson_contents: &str) {
        let mut url: String = String::new();
        if let hyper::uri::RequestUri::AbsolutePath(ref u) = req.uri {
            url = u.clone();
        }
            
        // FIXME use prefix
        match parse_url(&url, 22) {
            URL::Tilejson(_) => {
                tilejson_handler(res, tilejson_contents);
            },
            URL::Invalid => {
                *res.status_mut() = hyper::status::StatusCode::NotFound;
            },
            URL::Tile(_, z, x, y, _) => {
                tile_handler(res, tc_path, z, x, y, upstream_url);
            }
        }
    }

    println!("Serving on port {}", port);
    match Server::http(&*format!("localhost:{}", port)) {
        Err(e) => { println!("Couldn't open port: {:?}", e); return },
        Ok(s) => {
            s.handle(move |req: Request, res: Response| { base_handler(req, res, &tc_path, &upstream_url, &new_tilejson_contents) }).ok();
        },
    };
}


src/expire.rs
extern crate clap;
extern crate hyper;
extern crate slippy_map_tiles;
extern crate simple_parallel;
extern crate iter_progress;

use std::fs;
use std::path::{Path, PathBuf};
use std::io::BufReader;
use std::io::prelude::*;
use std::os::unix::fs::MetadataExt;
#[allow(deprecated)]
use std::os::unix::raw::time_t;
use std::thread::sleep;
use std::time::Duration;

use clap::ArgMatches;

use slippy_map_tiles::Tile;
use iter_progress::ProgressableIter;

use utils::download_url_and_save_to_file;

#[allow(deprecated)]
fn dl_tile_if_older(tile: Tile, tc_path: &str, upstream_url: &str, expiry_mtime: time_t) {
    let x = tile.x();
    let y = tile.y();
    let z = tile.zoom();

    let path = format!("{}/{}", tc_path, tile.tc_path("pbf"));
    let this_tile_tc_path = Path::new(&path);

    let should_dl = if ! this_tile_tc_path.exists() {
            true
        } else {
            let mtime = match this_tile_tc_path.metadata() {
                Err(e) => { println!("Error when trying to get tile metadata: {:?}", e); return; },
                Ok(m) => m.mtime(),
            };
            mtime < expiry_mtime
        };

    if should_dl {
        download_url_and_save_to_file(&format!("{}/{}/{}/{}.pbf", upstream_url, z, x, y), this_tile_tc_path).unwrap_or_else(|e| {
            println!("Error occured when downloading {}/{}/{}: {:?}", z, x, y, e);
        });
    }

}

fn get_expire_filenames(expire_directory: &Path) -> Result<Vec<PathBuf>, ()> {
    let entries = try!(expire_directory.read_dir().map_err(|_| ()));
    let entries = entries.filter_map(|entry| { entry.ok() });
    Ok(entries.filter(|entry| {
        let is_file = match entry.file_type().map(|f| f.is_file()) {
            Err(_) => { return false; }, Ok(x) => x };
        let file_name: String = match entry.path().file_name().and_then(|f| f.to_str() ) {
            None => { return false; } Some(x) => x.to_string() };

        is_file && file_name.starts_with("expire-") && file_name.ends_with(".txt")
    }).map(|entry| { entry.path() }).collect::<Vec<_>>())
}

fn single_expire_run(filename_path: &PathBuf, pool: &mut simple_parallel::Pool, tc_path: &str, upstream_url: &str) -> Result<(), String> {
    let filename = try!(try!(filename_path.file_name().ok_or("Couldn't get filename".to_string())).to_str().ok_or("Couldn't convert to string".to_string()));
    let file = try!(fs::File::open(&filename_path).map_err(|_| "Couldnt' open file".to_string()));
    let lines: Vec<_> = BufReader::new(file).lines().filter_map(|l| { l.ok() }).collect();
    println!("Processing {:?} which has {} lines", filename, lines.len());

    // Could use a .filter_map to get the from_tms, but we presume all the input is OK. Using a
    // map ensures that we have accurate sizes which means the iter-progress can give us
    // percentage views
    let tiles = lines.iter().map(|l| { Tile::from_tms(l.as_str()).unwrap() });

    let expiry_mtime = try!(filename_path.metadata().map_err(|_| "Couldn't get metadata".to_string())).mtime();

    pool.for_(tiles.progress(), |(state, tile)| {
        state.print_every_n_items(100, format!("{:.0}% done ({:.1}/sec), tile {:?}       \r", state.percent().map(|x| x.to_string()).unwrap_or("N/A".to_string()), state.rate(), tile));
        dl_tile_if_older(tile, &tc_path, &upstream_url, expiry_mtime);
    });
    let parent_dir = try!(filename_path.parent().ok_or("Directory".to_string()));
    let new_filename = &parent_dir.join(format!("done-{}", filename));


    fs::rename(&filename_path, new_filename).map_err(|_| "Couldn't rename".to_string())
}

pub fn expire(options: &ArgMatches) {

    let upstream_url = options.value_of("upstream_url").unwrap().to_string();
    let tc_path = options.value_of("tc_path").unwrap().to_string();
    let threads = options.value_of("threads").unwrap().parse().unwrap();

    let expire_path = options.value_of("expire_path").unwrap().to_string();

    let wait_between_runs = options.value_of("wait_between_runs").unwrap().parse().unwrap();


    println!("Starting {} threads", threads);
    let mut pool = simple_parallel::Pool::new(threads);

    let expire_directory = Path::new(&expire_path);



    loop {

        let expire_filenames = match get_expire_filenames(expire_directory) {
            Ok(e) => e,
            Err(_) => {
                // Something when wrong trying to get the files
                continue;
            },
        };

        if expire_filenames.len() == 0 {
            // Nothing to do, sleeping
            sleep(Duration::new(wait_between_runs, 0));
            continue;
        }

        println!("Found {} files ({:?}) to process", expire_filenames.len(), expire_filenames);

        for filename_path in expire_filenames {
            match single_expire_run(&filename_path, &mut pool, &tc_path, &upstream_url) {
                Ok(_) => {
                    println!("\nFinished processing file {:?}", filename_path);
                },
                Err(e) => {
                    println!("\nCouldn't download {:?} error: {:?}", filename_path, e);
                },
            }
        }
    }

}



src/stuffer.rs
extern crate clap;
extern crate hyper;
extern crate slippy_map_tiles;
extern crate simple_parallel;
extern crate iter_progress;
extern crate chrono;

use std::path::Path;
use std::os::unix::fs::MetadataExt;

use clap::ArgMatches;
use slippy_map_tiles::Tile;
use iter_progress::ProgressableIter;
use chrono::{DateTime, FixedOffset};

use utils::{download_url_and_save_to_file, IompairError, DirectoryLayout};

fn dl_tile(tile: Tile, path: &str, path_format: DirectoryLayout, upstream_url: &str, always_download: bool, files_older_than: &Option<DateTime<FixedOffset>>) -> Result<(), IompairError> {
    let x = tile.x();
    let y = tile.y();
    let z = tile.zoom();

    let this_path = format!("{}/{}", path, match path_format {
        DirectoryLayout::TCPath => tile.tc_path("pbf"),
        DirectoryLayout::TSPath => tile.ts_path("pbf"),
        DirectoryLayout::ZXYPath => tile.zxy_path("pbf"),
    });
    let this_path = Path::new(&this_path);

    let should_download = if ! this_path.exists() {
        true
    } else {
        if always_download {
            match *files_older_than {
                None => { true },
                Some(dt) => {
                    let mtime = this_path.metadata().unwrap().mtime();
                    let cutoff = dt.timestamp();
                    mtime < cutoff
                }
            }
        } else {
            false
        }
    };

    if should_download {
        try!(download_url_and_save_to_file(&format!("{}/{}/{}/{}.pbf", upstream_url, z, x, y), this_path));
    }

    Ok(())
}

fn dl_tilejson(path: &str, upstream_url: &str) -> Result<(), IompairError> {
    try!(download_url_and_save_to_file(&format!("{}/index.json", upstream_url), Path::new(&format!("{}/index.json", path))));
    Ok(())
}


pub fn stuffer(options: &ArgMatches) {

    let upstream_url = options.value_of("upstream_url").unwrap().to_string();
    let path = options.value_of("tc_path").or(options.value_of("ts_path")).or(options.value_of("zxy_path")).unwrap().to_string();
    let path_format = if options.is_present("tc_path") { DirectoryLayout::TCPath } else if options.is_present("ts_path") { DirectoryLayout::TSPath } else if options.is_present("zxy_path") { DirectoryLayout::ZXYPath } else { unreachable!() };
    let threads = options.value_of("threads").unwrap().parse().unwrap();
    let max_zoom = options.value_of("max-zoom").unwrap().parse().unwrap();
    let min_zoom: u8 = options.value_of("min-zoom").unwrap().parse().unwrap();

    let always_download = options.is_present("always-download");
    let files_older_than: Option<DateTime<FixedOffset>> = options.value_of("files-older-than").and_then(|t| { DateTime::parse_from_rfc3339(t).ok() });

    let top = options.value_of("top").unwrap().parse().unwrap();
    let bottom = options.value_of("bottom").unwrap().parse().unwrap();
    let left = options.value_of("left").unwrap().parse().unwrap();
    let right = options.value_of("right").unwrap().parse().unwrap();


    // Download the tilejson file and save it for later.
    dl_tilejson(&path, &upstream_url).unwrap_or_else(|e| {
        println!("Error occured when downloading tilejson: {:?}", e);
        println!("Aborting");
        return;
    });
    println!("Downloaded TileJSON");

    println!("Starting {} threads", threads);
    let mut pool = simple_parallel::Pool::new(threads);

    // FIXME unfortunate duplicate with the pool.for_ line

    if top == 90. && bottom == -90. && left == -180. && right == 180. {
        // We're doing the whole world
        let iter = Box::new(Tile::all_to_zoom(max_zoom).filter(|&t| { t.zoom() >= min_zoom }));
        pool.for_(iter.progress(), |(state, tile)| {
            state.print_every_n_sec(5., format!("{} done ({}/sec), tile {:?}       \r", state.num_done(), state.rate(), tile));
            dl_tile(tile, &path, path_format, &upstream_url, always_download, &files_older_than).unwrap_or_else(|e| {
                println!("Error occured when downloading tile {:?}: {:?}", tile, e);
            });
        });
    } else {
        match slippy_map_tiles::BBox::new(top, left, bottom, right) {
            None => {
                println!("Invalid bbox");
                return;
            },
            Some(b) => {
                let iter = b.tiles().filter(|&t| { t.zoom() >= min_zoom }).take_while(|&t| { t.zoom() <= max_zoom });

                pool.for_(iter.progress(), |(state, tile)| {
                    state.print_every_n_sec(1., format!("{} done ({}/sec), tile {:?}       \r", state.num_done(), state.rate(), tile));
                    //state.print_every_sec(100., format!("{} done ({}/sec), tile {:?}       \r", state.num_done(), state.rate(), tile));
                    dl_tile(tile, &path, path_format, &upstream_url, always_download, &files_older_than).unwrap_or_else(|e| {
                        println!("Error occured when downloading tile {:?}: {:?}", tile, e);
                    });
                });
            },
        }
    }

    print!("\n");
}


src/tilelist.rs
extern crate slippy_map_tiles;

use std::path::Path;
use clap::ArgMatches;
use slippy_map_tiles::Tile;

pub fn tilelist(options: &ArgMatches) {
    let max_zoom; let min_zoom;
    if options.is_present("zoom") {
        let zoom: u8 = options.value_of("zoom").unwrap().parse().unwrap();
        min_zoom = zoom;
        max_zoom = zoom;
    } else {
        min_zoom = options.value_of("min-zoom").unwrap_or("0").parse().unwrap();
        max_zoom = options.value_of("max-zoom").unwrap_or("14").parse().unwrap();
    }
    let not_exists = options.is_present("not_exists");
    let root_dir = options.value_of("ts_path");

    // FIXME iterating over all zoom levels less than min-zoom is probably a bit ineffecient, but
    // it's simple and it works

    for tile in Tile::all().take_while(|t| t.zoom() <= max_zoom).filter(|t| t.zoom() >= min_zoom) {
        let include = if not_exists {
            let tile_path = format!("{}/{}", root_dir.unwrap(), tile.ts_path("pbf"));
            let tile_path = Path::new(&tile_path);
            ! tile_path.exists()
        } else {
            // include everything
            true
        };

        if include {
            println!("{}/{}/{}", tile.zoom(), tile.x(), tile.y());
        }

    }
}


src/main.rs
extern crate hyper;
#[macro_use]
extern crate clap;
extern crate rustc_serialize;
extern crate slippy_map_tiles;
extern crate regex;
extern crate simple_parallel;
extern crate iter_progress;
extern crate chrono;
extern crate libflate;

use clap::{Arg, App, SubCommand, ArgGroup};


#[macro_use]
mod utils;

mod serve;
mod stuffer;
mod expire;
mod tilelist;

use serve::serve;
use stuffer::stuffer;
use expire::expire;
use tilelist::tilelist;

fn main() {

    let options = App::new("iompair")
        .about("Work with vector tiles")
        .version(crate_version!())
        .subcommand(SubCommand::with_name("serve")
            .about("Serve a tile cache directory")
            .arg(Arg::with_name("port").short("p").long("port")
                 .takes_value(true).required(true)
                 .help("Port to listen on").value_name("PORT"))
            .arg(Arg::with_name("maxzoom").short("z").long("max-zoom")
                 .takes_value(true).default_value("14")
                 .help("Maximum zoom to pretend").value_name("ZOOM"))
            .arg(Arg::with_name("urlprefix").long("urlprefix")
                 .takes_value(true).required(false)
                 .help("URL that the tiles are accessible under").value_name("URL"))
            .arg(Arg::with_name("tc_path").long("tc-path")
                 .takes_value(true).conflicts_with("ts_path")
                 .help("Directory to use as a tile cache (TileCache layout).").value_name("PATH"))
            .arg(Arg::with_name("ts_path").long("ts-path")
                 .takes_value(true).conflicts_with("tc-path")
                 .help("Directory to use as a tile cache (TileStash safe layout).").value_name("PATH"))
            .arg(Arg::with_name("zxy_path").long("zxy-path")
                 .takes_value(true).conflicts_with("tc-path").conflicts_with("ts-path")
                 .help("Directory to use as a tile cache (ZXY layout).").value_name("PATH"))
            .arg(Arg::with_name("verbose").long("verbose")
                 .takes_value(false)
                 .help("Verbose mode. Prints to stdout at every request served"))
            .group(ArgGroup::with_name("path").args(&["tc_path", "ts_path", "zxy_path"]).required(true))
            .arg(Arg::with_name("upstream_url").short("u").long("upstream")
                 .takes_value(true).multiple(true).number_of_values(2)
                 .help("Local prefix & the URL of the upstream vector tiles producer(s)").value_name("PREFIX URL"))
            .arg(Arg::with_name("post-fetch-command").long("post-fetch-command")
                 .takes_value(true).required(false).requires("upstream_url")
                 .help("When a tile has been downloaded from upstream, execute this command on it").value_name("COMMAND"))
            )
        .subcommand(SubCommand::with_name("stuffer")
            .about("Populate a tile cache directory with all the tiles in an area")
            .setting(clap::AppSettings::AllowLeadingHyphen)
            .arg(Arg::with_name("upstream_url").short("u").long("upstream")
                 .takes_value(true).required(true)
                 .help("URL of the upstream vector tiles producer").value_name("URL"))
            .arg(Arg::with_name("tc_path").short("c").long("tc-path")
                 .takes_value(true)
                 .help("Directory to use as a tile cache.").value_name("PATH"))
            .arg(Arg::with_name("ts_path").long("ts-path")
                 .takes_value(true).conflicts_with("tc-path")
                 .help("Directory to use as a tile cache (TileStash safe layout).").value_name("PATH"))
            .arg(Arg::with_name("threads").short("T").long("threads")
                 .takes_value(true).required(false).default_value("4")
                 .help("Number of threads").value_name("THREADS"))
            .arg(Arg::with_name("max-zoom").short("z").long("max-zoom")
                 .takes_value(true).required(false).default_value("14")
                 .help("Maximum zoom to go to").value_name("ZOOM"))
            .arg(Arg::with_name("min-zoom").long("min-zoom")
                 .takes_value(true).required(false).default_value("0")
                 .help("Minimum zoom to start from").value_name("ZOOM"))
            .arg(Arg::with_name("top").short("t").long("top")
                 .takes_value(true).required(false).default_value("90"))
            .arg(Arg::with_name("left").short("l").long("left")
                 .takes_value(true).required(false).default_value("-180"))
            .arg(Arg::with_name("bottom").short("b").long("bottom")
                 .takes_value(true).required(false).default_value("-90"))
            .arg(Arg::with_name("right").short("r").long("right")
                 .takes_value(true).required(false).default_value("180"))
            .arg(Arg::with_name("always-download").long("always-download")
                 .takes_value(false).required(false)
                 .help("Always download the files, even if they already exist"))
            .arg(Arg::with_name("files-older-than").long("files-older-than")
                 .takes_value(true).required(false)
                 .help("If using --always-download, only download a file that's missing or older than this RFC3339 datetime"))
            )
        .subcommand(SubCommand::with_name("expire")
            .about("Update a tilecache directory from upstream with osm2pgsql expiry tile list")
            .arg(Arg::with_name("upstream_url").short("u").long("upstream")
                 .takes_value(true).required(true)
                 .help("URL of the upstream vector tiles producer").value_name("URL"))
            .arg(Arg::with_name("tc_path").short("c").long("tc-path")
                 .takes_value(true).required(true)
                 .help("Directory to use as a tile cache.").value_name("PATH"))
            .arg(Arg::with_name("threads").short("T").long("threads")
                 .takes_value(true).required(false).default_value("4")
                 .help("Number of threads").value_name("THREADS"))
            .arg(Arg::with_name("expire_path").short("e").long("expire-path")
                 .takes_value(true).required(true)
                 .help("Directory which stores the expire-*.txt files").value_name("PATH"))
            .arg(Arg::with_name("wait_between_runs").short("w").long("wait")
                 .takes_value(true).required(false).default_value("60")
                 .help("How long (in SEC) to wait between checks of the expire directory. Default 60 sec").value_name("SEC"))
            )
        .subcommand(SubCommand::with_name("tilelist")
            .about("Generate a Z/X/Y tile list (to stdout) based on tiles")
            .arg(Arg::with_name("max-zoom").short("z").long("max-zoom")
                 .takes_value(true).required(false)
                 .help("Maximum zoom to go to").value_name("ZOOM"))
            .arg(Arg::with_name("min-zoom").long("min-zoom")
                 .takes_value(true).required(false)
                 .help("Minimum zoom to start from").value_name("ZOOM"))
            .arg(Arg::with_name("not_exists").long("--not-exists")
                 .takes_value(false).required(false)
                 .help("Only include files which don't exist"))
            .arg(Arg::with_name("zoom").long("zoom")
                 .takes_value(true).required(false)
                 .help("Only generate tiles on this zoom level").value_name("ZOOM")
                 .conflicts_with("min-zoom").conflicts_with("max-zoom")
                 )
            .arg(Arg::with_name("ts_path").long("ts-path")
                 .takes_value(true)
                 .help("Directory to use as a tile cache (TileStash safe layout).").value_name("PATH"))
            )
        .get_matches();

    match options.subcommand() {
        ("serve", Some(options)) => { serve(options); },
        ("stuffer", Some(options)) => { stuffer(options); },
        ("expire", Some(options)) => { expire(options); },
        ("tilelist", Some(options)) => { tilelist(options); },
        (_, _) => { println!("{}", options.usage()); },
    }

}
